### PR TITLE
Snapshot support

### LINKED ISSUE #47: Snapshotting and rewinding
For a project I could really use an API that allows me to rollback to a snapshot of a tome. I'm thinking it could be as simple as this:

```js
Tome.prototype.takeSnapshot = function () {
	if (this.__root__ !== this) {
		throw new Error('Only root tomes can take snapshots');
	}

	return {
		content: Tome.unTome(this),
		diff: this.__diff__.slice()
	};
};


Tome.prototype.restoreSnapshot = function (snapshot) {
	if (this.__root__ !== this) {
		throw new Error('Only root tomes can restore snapshots');
	}

	this.__version__ = 1;
	this.__diff__ = snapshot.diff.slice();
	this.assign(snapshot.content, { silent: true });
};
```

The version would be reset to 1, but it doesn't seem to me like that would be a problem. The `silent: true` option would ensure that during the assign operation, no diffs are created (thus the version will remain 1) and nothing gets emitted. Because the whole tome gets reassigned a value without diffs, all the dirty flags would be set to 1, so it would equal the version, just like right after a conjure.

The reason why this would this be a useful API (rather than just making a snapshot from the outside and conjuring it) is twofold:

- diffs are restored to where they were at the snapshot point-in-time.
- any outside reference to the root tome remains valid.

Sidenote:

If we want to keep the version as-is, we could store the version in the snapshot, and when we conjure we don't initialize `__dirty__` to `1`, but instead set it to `__root__.__version__` (which is arguably more correct anyway). Doing this, we could also optimize the restoreSnapshot method to be a no-op if there's no version change.

### PR DESCRIPTION
**Pull request by ronkorving**
Fixes #47

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -262,18 +266,21 @@ function markDirty(tome, dirtyAt, was) {
...
+		}
 
-	root.__diff__.push(newOp);
+		var newOp = { chain: chain, op: op };
+		if (val !== undefined) {
+			newOp.val = val;
+		}
+
+		root.__diff__.push(newOp);
+	}
```
Comment (ronkorving) [inline index.js]: The only thing that changed here was that a block of code was moved into an if-statement. Side-by-side diff will make this clear.

File: index.js [hunk 2]
```diff
@@ -631,6 +638,40 @@ Tome.prototype.getParent = function () {
...
+		// nothing would change from restoring
+		return;
+	}
+
+	this.__version__ = snapshot.version;
+	this.__diff__ = snapshot.diff.slice();
+
+	this.__diffEnabled__ = false;
+
+	try {
```
Comment (bjornstar) [inline index.js]: you're just swallowing errors here :scream:

Comment (ronkorving) [inline index.js]: No, note there is no catch clause.

File: index.js [hunk 3]
```diff
@@ -631,6 +638,40 @@ Tome.prototype.getParent = function () {
...
+		this.assign(snapshot.content);
```
Comment (AlmirKadric) [inline index.js]: won't this generate diffs and version bumps?

File: index.js [hunk 4]
```diff
@@ -84,7 +87,7 @@ function Tome(parent, key) {
 	var hasParentTome = Tome.isTome(parent);
 
 	var properties = {
-		__dirty__: { writable: true, value: 1 },
+		__dirty__: { writable: true, value: hasParentTome ? parent.__root__.__version__ : 1 },
```
Comment (AlmirKadric) [inline index.js]: why do we carry the parent version into the children???

**Comment by maxwan** [comment]
LGTM

File: index.js (hunk 4 above)
Comment (ronkorving) [inline index.js]: Read all of #47 for full context please

**Comment by ronkorving** [comment]
If there are no more objections, I would like to merge this tomorrow.

**Comment by AlmirKadric** [comment]
LGTM :+1:

File: index.js (hunk 3 above)
Comment (ronkorving) [inline index.js]: I guess technically, because this is synchronous, that flag could be system-wide, it doesn't have to sit on the tome at all.

Comment (ronkorving) [inline index.js]: Ah, but the problem with that is that we do emit events during `markDirty`, and if anyone would ever use this event to start restoring other snapshots this global would be screwed. So nevermind.

File: index.js (hunk 4 above)
Comment (ronkorving) [inline index.js]: Did this make sense?

Comment (AlmirKadric) [inline index.js]: yeah it made sense

**Comment by ronkorving** [comment]
Then boom, landing it.

**Pull request merged**